src/cli/test/Scanner.zig

```diff
@@ -0,0 +1,209 @@
+const Scanner = @This();
+
+const std = @import("std");
+const bun = @import("root").bun;
+const BundleOptions = @import("../../options.zig").BundleOptions;
+
+const Allocator = std.mem.Allocator;
+const FileSystem = bun.fs.FileSystem;
+const Transpiler = bun.Transpiler;
+const strings = bun.strings;
+const StringOrTinyString = strings.StringOrTinyString;
+const JSC = bun.JSC;
+const jest = JSC.Jest;
+
+/// Memory is borrowed.
+exclusion_names: []const []const u8 = &.{},
+/// When no filters exist, none are applied. Memory is borrowed.
+filter_names: []const []const u8 = &.{},
+dirs_to_scan: Fifo,
+/// Scanner mutably borrows this results list
+results: *std.ArrayList(bun.PathString),
+fs: *FileSystem,
+open_dir_buf: bun.PathBuffer = undefined,
+scan_dir_buf: bun.PathBuffer = undefined,
+options: *BundleOptions,
+has_iterated: bool = false,
+search_count: usize = 0,
+
+const Fifo = std.fifo.LinearFifo(ScanEntry, .Dynamic);
+const ScanEntry = struct {
+    relative_dir: bun.StoredFileDescriptorType,
+    dir_path: []const u8,
+    name: StringOrTinyString,
+};
+const Error = error{
+    /// Scan entrypoint file/directory does not exist. Not returned when
+    /// a subdirectory is scanned but does not exist.
+    DoesNotExist,
+} || Allocator.Error;
+
+pub fn init(
+    allocator: Allocator,
+    transpiler: *Transpiler,
+    results: *std.ArrayList(bun.PathString),
+) Scanner {
+    return Scanner{
+        .dirs_to_scan = Fifo.init(allocator),
+        .options = &transpiler.options,
+        .fs = transpiler.fs,
+        .results = results,
+    };
+}
+
+pub fn deinit(this: *Scanner) void {
+    this.dirs_to_scan.deinit();
+    // NOTE: results is borrowed
+    this.* = undefined;
+}
+
+fn readDirWithName(this: *Scanner, name: []const u8, handle: ?std.fs.Dir) !*FileSystem.RealFS.EntriesOption {
+    return try this.fs.fs.readDirectoryWithIterator(name, handle, 0, true, *Scanner, this);
+}
+
+pub fn scan(this: *Scanner, path_literal: []const u8) Error!void {
+    const parts = &[_][]const u8{ this.fs.top_level_dir, path_literal };
+    const path = this.fs.absBuf(parts, &this.scan_dir_buf);
+
+    var root = this.readDirWithName(path, null) catch |err| {
+        switch (err) {
+            error.NotDir, error.ENOTDIR => {
+                if (this.isTestFile(path)) {
+                    this.results.append(bun.PathString.init(this.fs.filename_store.append([]const u8, path) catch bun.outOfMemory())) catch bun.outOfMemory();
+                }
+            },
+            error.ENOENT => return error.DoesNotExist,
+            error.OutOfMemory => return error.OutOfMemory,
+            else => std.debug.print("error: {s}\n", .{@errorName(err)}),
```

Comment:
please don't use stub.debug.print

Reply:
std actually haha

Reply:
oops, had that there for debugging 👀 removed